components/script/script_module.rs

```diff
@@ -652,24 +652,105 @@ impl ModuleTree {
+        // Step 6. Let serializedBaseURL be baseURL, serialized.
+        let base_url = base_url.as_str();
+        // Step 8. Let normalizedSpecifier be the serialization of asURL, if asURL is non-null;
+        // otherwise, specifier.
+        let normalized_specifier = match &as_url {
+            Some(url) => url.as_str(),
+            None => &specifier,
+        };
+
+        // Step 9. Let result be a URL-or-null, initially null.
+        let mut result = None;
+        if let Some(map) = import_map {
+            // Step 10. For each scopePrefix → scopeImports of importMap's scopes:
+            for (prefix, imports) in &map.scopes {
+                // Step 10.1 If scopePrefix is serializedBaseURL, or if scopePrefix ends with U+002F (/)
+                // and scopePrefix is a code unit prefix of serializedBaseURL, then:
+                let prefix = prefix.as_str();
+                if prefix == base_url ||
+                    (base_url.starts_with(prefix) && prefix.ends_with('\u{002f}'))
+                {
+                    // Step 10.1.1 Let scopeImportsMatch be the result of resolving an imports match
+                    // given normalizedSpecifier, asURL, and scopeImports.
+                    // Step 10.1.2 If scopeImportsMatch is not null, then set result to scopeImportsMatch,
+                    // and break.
+                    result = resolve_imports_match(
+                        normalized_specifier,
+                        as_url.as_ref(),
+                        imports,
+                        can_gc,
+                    )?;
+                    break;
+                }
+            }
+
+            // Step 11. If result is null, set result to the result of resolving an imports match given
+            // normalizedSpecifier, asURL, and importMap's imports.
+            if result.is_none() {
+                result = resolve_imports_match(
+                    normalized_specifier,
+                    as_url.as_ref(),
+                    &map.imports,
+                    can_gc,
+                )?;
+            }
+        }
+
+        // Step 12. If result is null, set it to asURL.
+        if result.is_none() {
+            result = as_url.clone();
+        }
+
+        // Step 13. If result is not null, then:
+        match result {
+            Some(result) => {
+                // Step 13.1 Add module to resolved module set given settingsObject, serializedBaseURL,
+                // normalizedSpecifier, and asURL.
+                // <https://html.spec.whatwg.org/multipage/#add-module-to-resolved-module-set>
+                if global.is::<Window>() {
+                    let record = ResolvedModule {
+                        base_url: base_url.to_owned(),
+                        specifier: normalized_specifier.to_owned(),
+                        specifier_url: as_url,
+                    };
+                    global.resolved_module_set().borrow_mut().insert(record);
```

Comment:
I would add a method to `GlobalScope` that implements https://html.spec.whatwg.org/multipage/webappapis.html#add-module-to-resolved-module-set